### PR TITLE
add project choose for Talisman

### PR DESCRIPTION
**Pull request by Hat331**

### CODE REVIEW

File: src/scripts/StoreTalismanApi.js [hunk 1]
```diff
@@ -72,4 +83,111 @@ export default class StoreTalismanApi extends StoreRestApi {
 		delete this.axiosInstance.defaults.headers.Authorization;
 		await this.axiosInstance.get('/oauth/logout');
 	}
+
+	async getAllProjects() {
+		this.setSitemapPath();
+		const queryData = {
+			query: GET_ALL_PROJECTS_QUERY,
+			variables: {
+				limit: 100,
```
Comment (mxsnq) [inline src/scripts/StoreTalismanApi.js:92]: move to constant and increase

File: src/scripts/StoreTalismanApi.js [hunk 2]
```diff
@@ -72,4 +83,111 @@ export default class StoreTalismanApi extends StoreRestApi {
...
+				},
+				sortField: 'title',
+				sortDirection: 'ascending',
+			},
+		};
+		const sitemaps_obj = await this.axiosInstance.post('/graphql', queryData);
+		return sitemaps_obj.data.data.paginationCrawler.listCrawler.map(sitemap =>
+			JSON.parse(sitemap.sitemap)
+		);
+	}
```
Comment (mxsnq) [inline src/scripts/StoreTalismanApi.js:120]: - why gql?
- where is limit
- restore sitemap migrations

File: src/scripts/StoreTalismanApi.js [hunk 3]
```diff
@@ -72,4 +83,111 @@ export default class StoreTalismanApi extends StoreRestApi {
...
+	deleteSitemap(sitemap) {
+		return this.axiosInstance
+			.delete(urlJoin(this.sitemapsPath, sitemap._id))
+			.then(response => {
+				return response.data;
+			})
+			.catch(() => {
+				alert('StoreApi: Error deleting sitemap.');
+			});
+	}
```
Comment (mxsnq) [inline src/scripts/StoreTalismanApi.js:161]: copies rest api storage

File: src/scripts/StoreTalismanApi.js [hunk 4]
```diff
@@ -72,4 +83,111 @@ export default class StoreTalismanApi extends StoreRestApi {
...
+			},
+		};
+		return this.axiosInstance.post('/graphql', queryData);
+	}
+
+	async getAllSitemaps(projectId) {
+		if (!projectId) {
+			projectId = this.sitemapsPath.split('/')[2];
+		}
+		this.setSitemapPath(projectId);
```
Comment (mxsnq) [inline src/scripts/StoreTalismanApi.js:104]: Do not implicitly set project while accessing its sitemaps. Either set project explicitly from controller or make store methods accept project argument.

File: src/devtools/views/ProjectListItem.html [hunk 5]
```diff
@@ -0,0 +1,4 @@
+<tr>
+	<td class="id">{{id}}</td>
+	<td>{{#title}} {{title}} {{/title}}</td>
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/devtools/views/ProjectListItem.html]: fix

File: src/scripts/Controller.js [hunk 6]
```diff
@@ -226,6 +230,9 @@ export default class SitemapController {
 			'#sitemaps tr td:nth-of-type(1)': {
 				click: this.editSitemap,
 			},
+			'#projects tr td:nth-of-type(1)': {
+				click: this.editProject,
+			},
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/Controller.js]: fix

File: src/devtools/views/Viewport.html [hunk 7]
```diff
@@ -17,7 +17,9 @@
 </style>
 <nav class="navbar navbar-default disabled" role="navigation">
 	<ul class="nav navbar-nav">
-		<li><a id="sitemaps-nav-button" data-i18n="sitemaps_nav_button"></a></li>
+		<li><a id="projects-nav-button" data-i18n="projects_nav_button"></a></li>
+		<!-- RENAME i18-n -->
+		<li><a id="sitemaps-nav-button" class="disabled" data-i18n="sitemaps_nav_button"></a></li>
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/devtools/views/Viewport.html]: fix

File: src/scripts/StoreTalismanApi.js (hunk 1 above)
Comment (Hat331) [reply src/scripts/StoreTalismanApi.js]: fix

File: src/scripts/StoreTalismanApi.js (hunk 2 above)
Comment (Hat331) [reply src/scripts/StoreTalismanApi.js]: fix

File: src/scripts/StoreTalismanApi.js (hunk 3 above)
Comment (Hat331) [reply src/scripts/StoreTalismanApi.js]: fix

File: src/devtools/views/Viewport.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (mxsnq) [reply src/devtools/views/Viewport.html]: use project.title instead of name

File: src/scripts/Controller.js [hunk 8]
```diff
@@ -719,13 +730,57 @@ export default class SitemapController {
...
+	async showSitemaps(projectId, projectName) {
 		this.clearState();
-		this.setActiveNavigationButton('sitemaps');
 
-		const sitemaps = await this.store.getAllSitemaps();
+		if (typeof projectId === 'string') {
+			$('#sitemaps-nav-button span#navbar-active-project-id').text(`(${projectName})`);
+			$(`#sitemaps-nav-button`).attr('projectid', projectId);
+		} else {
+			projectId = $(`#sitemaps-nav-button`).attr('projectid');
```
Comment (mxsnq) [inline src/scripts/Controller.js:774]: Move project id extraction into method similar to getCurrentlyEditedXXX, or populate controller state with it.

File: src/devtools/views/AuthorizationPage.html [hunk 9]
```diff
@@ -1,3 +1,10 @@
+<!--<style>-->
```
Comment (mxsnq) [inline src/devtools/views/AuthorizationPage.html:1]: New auth panel looks awesome!
Last thing, please tidy up the file and remove commented out code.

**Review by mxsnq — CHANGES_REQUESTED** [review]
There are several methods in controller where project id is missing, e.g.
- on sitemap import
- on sitemap copy
- when editing sitemap metadata

File: src/devtools/views/Viewport.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (Hat331) [reply src/devtools/views/Viewport.html]: done

File: src/scripts/Controller.js (hunk 8 above)
Comment (Hat331) [reply src/scripts/Controller.js]: done

File: src/devtools/views/AuthorizationPage.html (hunk 9 above)
Comment (Hat331) [reply src/devtools/views/AuthorizationPage.html]: done

**Review by mxsnq — COMMENTED** [review]
Potential bug: sitemaps with same id in different projects will share common local items store.

**Comment by mxsnq** [comment]
Don't forget to update version in package.json

File: src/scripts/Controller.js [hunk 10]
```diff
@@ -719,13 +733,57 @@ export default class SitemapController {
...
+
+		const projects = await this.store.getAllProjects();
+
+		if (projects.error_msg) {
+			$('#sitemaps').hide();
+			$('#viewport').html(
+				'<div class="container"><div data-i18n="get_sitemap_error"></div></div>'
+			);
+			Translator.translatePage();
+		}
```
Comment (mxsnq) [inline src/scripts/Controller.js:751]: I believe we don't send error_msg from storage, neither does gql api

File: src/scripts/TalismanStoreDevtools.js [hunk 11]
```diff
@@ -28,4 +29,62 @@ export default class TalismanStoreDevtools extends StoreDevtools {
...
+		return this._getAllSitemapsResponseHandler(await browser.runtime.sendMessage(request));
+	}
+
+	async getAllProjects() {
+		const request = {
+			getAllProjects: true,
+		};
+		return await browser.runtime
+			.sendMessage(request)
+			.then(res => res.data.data.paginationProject.listProject);
```
Comment (mxsnq) [inline src/scripts/TalismanStoreDevtools.js:47]: I'd move this `.data.data.paginationProject.listProject` to StoreTalismanApi, where the query is

File: src/scripts/Controller.js (hunk 10 above)
Comment (Hat331) [reply src/scripts/Controller.js]: done

File: src/scripts/Controller.js [hunk 12]
```diff
@@ -18,6 +18,9 @@ import Translator from './Translator';
 
 export default class SitemapController {
 	constructor(store, templateDir) {
+		this.currentProjectId = () => {
+			return $(`#sitemaps-nav-button`).attr('projectid');
+		};
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/Controller.js]: done

File: src/scripts/Controller.js [hunk 13]
```diff
@@ -1255,7 +1323,7 @@ export default class SitemapController {
 		}
 		try {
 			sitemap.updateSelector(selector, newSelector);
-			await this.store.saveSitemap(sitemap);
+			await this.store.saveSitemap(sitemap, this.currentProjectId());
```
[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/Controller.js]: done

[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/Controller.js]: > Potential bug: sitemaps with same id in different projects will share common local items store.

Fixed

[the remark this answers is not in the document]
Comment (Hat331) [reply src/scripts/Controller.js]: > Don't forget to update version in package.json

done

**Review by mxsnq — APPROVED** [review]

**Pull request merged**